### DISCUSSION TITLE
How to pass data to a solver ?

### DISCUSSION THREAD

**Discussion by avhz** [argmin-rs/argmin#484] [Q&A]
If I want to minimise a squared error (or similar) between a function and some data, how do I pass the data to the cost function ?

**Comment by stefan-k** [comment]
You can do this via your problem definition:

```rust
struct MyProblem {
    data: MyData,
}

impl CostFunction for MyProblem {
    type Param = Vec<f64>;
    type Output = f64;

    fn cost(&self, p: &Self::Param) -> Result<Self::Output, Error> {
        Ok(my_cost_function(p, self.data))
    }
}
```

**Comment by avhz** [comment]
Is it possible to have a model defined by a struct that contains the parameters, and then solve for those parameters against some exogenous data ?

Something like:

```rust
struct MyModel {
    a: f64,
    b: f64,
}

impl MyModel {
    fn my_line(&self, x: f64) -> f64 {
        self.a * x + self.b
    }

    fn my_cost(&self, x: f64, y: f64) -> f64 {
        (self.my_line(x) - y).powi(2)
    }
}

const SOME_X_DATA: [f64; 9] = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0];
const SOME_Y_DATA: [f64; 9] = [0.0, 1.0, 2.0, 4.0, 7.0, 9.0, 5.0, 3.0, 7.0];
```

I would expect a result of: Y ~= 0.75 * X + 0.4722

So, I want to get:

```rust
MyModel { 
    a: 0.75, 
    b: 0.4722 
}
```

Thanks for any advice you can give ! Cheers!

**Comment by stefan-k** [comment]
Do you mean such that `type Param = MyModel` in the `CostFunction` trait impl? Yes, this is possible, but it will require you to implement several math related types on `MyModel` (which of those traits depends on the solver you want to use). I would suggest modelling your problem based on `Vec<f64>` or the equivalent ndarray/nalgebra types and then implement the conversion from `Vec<f64>` to `MyModel`:

```rust
struct MyProblem {}

impl CostFunction for MyProblem {
    type Param = Vec<f64>;
    type Output = f64;

    fn cost(&self, p: &Self::Param) -> Result<Self::Output, Error> {
        let cost = (0..9).iter().map(|i| (p[0] * SOME_X_DATA[i] + p[1] - SOME_Y_DATA[i]).powi(2)).sum();
        Ok(cost)
    }
}

// Then run optimization, obtain final solution `res`
// ...

let final_params = MyModel { a: res[0], b: res[1] }; // Of course  you can also implement `From<Vec<f64>>` for `MyModel` 
```

Then `a = res[0]` and `b = res[1]` where `res` is the final solution.

**Reply by avhz** [reply]
This is super helpful, thank you !

**Comment by avhz** [comment]
Hi, taking a look at this again after a while. 

If the additional data/arguments are not `const` as I gave previously, how can they be passed to the solver?

e.g. `scipy.optimize.root(fun, x0, args=(), ...)` has the `args` argument for this purpose. How can I achieve this if the only input to `cost` is the parameter vector ?

**Reply by stefan-k** [reply]
Does the data change during an optimization run?

**Reply by avhz** [reply]
No, just extra arguments to the function.

**Reply by stefan-k** [reply]
You can add those arguments as fields to your problem struct and set them when creating an instance of your problem. Then you have them available in your cost function.
